server/db/db.js

```diff
@@ -67,19 +67,21 @@ async function beginTransaction(retryCount = 0) {
 
     await pgClient.query('LOCK TABLE clients, sessions, locations')
   } catch (e) {
-    if (retryCount < 1) {
-      helpers.logError(`Error running the beginTransaction query: ${e}. Will retry beginTransaction again.`)
-      beginTransaction(retryCount + 1)
-    }
     helpers.logError(`Error running the beginTransaction query: ${e}`)
+
     if (pgClient) {
       try {
         await this.rollbackTransaction(pgClient)
-        return null
       } catch (err) {
         helpers.logError(`beginTransaction: Error rolling back the errored transaction: ${err}`)
       }
     }
+
+    if (retryCount < 1) {
+      helpers.log(`Retrying beginTransaction.`)
+      return await beginTransaction(retryCount + 1)
```

Comment:
By having the `retry` a parameter into this function, you're opening this function up to questions such as "What happens if you call `beginTransaction(3)`?"

From the outside of `db.js`, we only ever want to call `beginTransaction` with no parameters, right?

What do you think about splitting this up into two functions? Like this:
1. `beginTransaction` (which takes no parameters and is exported and accessible from outside db.js)
2. `runBeginTransactionWithRetries` or whatever you think would be a good name for it (which takes an integer `retry` parameter and is not accessible from outside db.js)

Then `beginTransaction` will be just `return await runBeginTransactionWithRetries(0)`

and you can rename what you've written here to be `runBeginTransactionWithRetries` instead.

Reply:
Oh this is a really good flag, that I hadn't thought of! I like the idea of splitting up the function into two. I'll give that a go.

Reply:
You see this sort of pattern quite often in languages like Java that allow you to reuse a function name with different parameters. Unfortunately you can't do that in JS, so the two functions need different names, which can be annoying. Oh well! Hopefully it makes things clearer.